# 1.) Import the Credit Card Fraud Data From CCLE

In [1]:
import pandas as pd
from google.colab import drive
import matplotlib.pyplot as plt
import numpy as np

In [2]:
drive.mount('/content/gdrive/', force_remount = True)

Mounted at /content/gdrive/


In [3]:
df = pd.read_csv("/content/gdrive/MyDrive/Econ441B/fraudTest.csv")

In [4]:
df.head()

Unnamed: 0 trans_date_trans_time            cc_num  \
0           0   2020-06-21 12:14:25  2291163933867244   
1           1   2020-06-21 12:14:33  3573030041201292   
2           2   2020-06-21 12:14:53  3598215285024754   
3           3   2020-06-21 12:15:15  3591919803438423   
4           4   2020-06-21 12:15:17  3526826139003047   

                               merchant        category    amt   first  \
0                 fraud_Kirlin and Sons   personal_care   2.86    Jeff   
1                  fraud_Sporer-Keebler   personal_care  29.84  Joanne   
2  fraud_Swaniawski, Nitzsche and Welch  health_fitness  41.28  Ashley   
3                     fraud_Haley Group        misc_pos  60.05   Brian   
4                 fraud_Johnston-Casper          travel   3.19  Nathan   

       last gender                       street  ...      lat      long  \
0   Elliott      M            351 Darlene Green  ...  33.9659  -80.9355   
1  Williams      F             3638 Marsh Union  ...  40.3207 -110.4360   
2     Lopez      F         9333 Valentine Point  ...  40.6729  -73.5365   
3  Williams      M  32941 Krystal Mill Apt. 552  ...  28.5697  -80.8191   
4    Massey      M     5783 Evan Roads Apt. 465  ...  44.2529  -85.0170   

   city_pop                     job         dob  \
0    333497     Mechanical engineer  1968-03-19   
1       302  Sales professional, IT  1990-01-17   
2     34496       Librarian, public  1970-10-21   
3     54767            Set designer  1987-07-25   
4      1126      Furniture designer  1955-07-06   

                          trans_num   unix_time  merch_lat  merch_long  \
0  2da90c7d74bd46a0caf3777415b3ebd3  1371816865  33.986391  -81.200714   
1  324cc204407e99f51b0d6ca0055005e7  1371816873  39.450498 -109.960431   
2  c81755dbbbea9d5c77f094348a7579be  1371816893  40.495810  -74.196111   
3  2159175b9efe66dc301f149d3d5abf8c  1371816915  28.812398  -80.883061   
4  57ff021bd3f328f8738bb535c302a31b  1371816917  44.959148  -85.884734   

   is_fraud  
0         0  
1         0  
2         0  
3         0  
4         0  

[5 rows x 23 columns]

In [6]:
df.columns

Index(['Unnamed: 0', 'trans_date_trans_time', 'cc_num', 'merchant', 'category',
       'amt', 'first', 'last', 'gender', 'street', 'city', 'state', 'zip',
       'lat', 'long', 'city_pop', 'job', 'dob', 'trans_num', 'unix_time',
       'merch_lat', 'merch_long', 'is_fraud'],
      dtype='object')

In [7]:
df.dtypes

Unnamed: 0                 int64
trans_date_trans_time     object
cc_num                     int64
merchant                  object
category                  object
amt                      float64
first                     object
last                      object
gender                    object
street                    object
city                      object
state                     object
zip                        int64
lat                      float64
long                     float64
city_pop                   int64
job                       object
dob                       object
trans_num                 object
unix_time                  int64
merch_lat                float64
merch_long               float64
is_fraud                   int64
dtype: object

In [5]:
df['dob'] = pd.to_datetime(df['dob'])

In [6]:
df['dob'].dt.year.head()

0    1968
1    1990
2    1970
3    1987
4    1955
Name: dob, dtype: int64

In [7]:
df['year_birth'] = df['dob'].dt.year

In [8]:
df['customer_classification'] = df['year_birth'].copy()
# df["customer_classification"] = df["customer_classification"].astype(int)

In [9]:
# ya = 0, midd = 1, eld = 2
df.loc[df['customer_classification'] < 1960, 'customer_classification'] = 2
df.loc[(df['customer_classification'] <= 1985) & (df['customer_classification']>= 1960), 'customer_classification'] = 1
df.loc[(df['customer_classification'] <= 2005) & (df['customer_classification']>= 1986), 'customer_classification'] = 0

In [10]:
df.loc[df['customer_classification'] == 2, 'customer_classification'] = 'elderly'
df.loc[df['customer_classification'] == 1, 'customer_classification'] = 'middle_aged'
df.loc[df['customer_classification'] == 0, 'customer_classification'] = 'young_adult'

I created a customer classification variable here based on the customers date of birth. My rational was that I wanted to see if certain age brackets, such as the elderly, got targeted more by fraudsters. This variable will be used in the model seen in the coming problems.

In [15]:
df

Unnamed: 0 trans_date_trans_time            cc_num  \
0                0   2020-06-21 12:14:25  2291163933867244   
1                1   2020-06-21 12:14:33  3573030041201292   
2                2   2020-06-21 12:14:53  3598215285024754   
3                3   2020-06-21 12:15:15  3591919803438423   
4                4   2020-06-21 12:15:17  3526826139003047   
...            ...                   ...               ...   
555714      555714   2020-12-31 23:59:07    30560609640617   
555715      555715   2020-12-31 23:59:09  3556613125071656   
555716      555716   2020-12-31 23:59:15  6011724471098086   
555717      555717   2020-12-31 23:59:24     4079773899158   
555718      555718   2020-12-31 23:59:34  4170689372027579   

                                    merchant        category     amt    first  \
0                      fraud_Kirlin and Sons   personal_care    2.86     Jeff   
1                       fraud_Sporer-Keebler   personal_care   29.84   Joanne   
2       fraud_Swaniawski, Nitzsche and Welch  health_fitness   41.28   Ashley   
3                          fraud_Haley Group        misc_pos   60.05    Brian   
4                      fraud_Johnston-Casper          travel    3.19   Nathan   
...                                      ...             ...     ...      ...   
555714                 fraud_Reilly and Sons  health_fitness   43.77  Michael   
555715                  fraud_Hoppe-Parisian       kids_pets  111.84     Jose   
555716                       fraud_Rau-Robel       kids_pets   86.88      Ann   
555717                 fraud_Breitenberg LLC          travel    7.99     Eric   
555718                     fraud_Dare-Marvin   entertainment   38.13   Samuel   

            last gender                       street  ... city_pop  \
0        Elliott      M            351 Darlene Green  ...   333497   
1       Williams      F             3638 Marsh Union  ...      302   
2          Lopez      F         9333 Valentine Point  ...    34496   
3       Williams      M  32941 Krystal Mill Apt. 552  ...    54767   
4         Massey      M     5783 Evan Roads Apt. 465  ...     1126   
...          ...    ...                          ...  ...      ...   
555714     Olson      M          558 Michael Estates  ...      519   
555715   Vasquez      M          572 Davis Mountains  ...    28739   
555716    Lawson      F   144 Evans Islands Apt. 683  ...     3684   
555717   Preston      M   7020 Doyle Stream Apt. 951  ...      129   
555718      Frey      M     830 Myers Plaza Apt. 384  ...   116001   

                           job        dob                         trans_num  \
0          Mechanical engineer 1968-03-19  2da90c7d74bd46a0caf3777415b3ebd3   
1       Sales professional, IT 1990-01-17  324cc204407e99f51b0d6ca0055005e7   
2            Librarian, public 1970-10-21  c81755dbbbea9d5c77f094348a7579be   
3                 Set designer 1987-07-25  2159175b9efe66dc301f149d3d5abf8c   
4           Furniture designer 1955-07-06  57ff021bd3f328f8738bb535c302a31b   
...                        ...        ...                               ...   
555714            Town planner 1966-02-13  9b1f753c79894c9f4b71f04581835ada   
555715          Futures trader 1999-12-27  2090647dac2c89a1d86c514c427f5b91   
555716                Musician 1981-11-29  6c5b7c8add471975aa0fec023b2e8408   
555717            Cartographer 1965-12-15  14392d723bb7737606b2700ac791b7aa   
555718             Media buyer 1993-05-10  1765bb45b3aa3224b4cdcb6e7a96cee3   

         unix_time  merch_lat  merch_long is_fraud year_birth  \
0       1371816865  33.986391  -81.200714        0       1968   
1       1371816873  39.450498 -109.960431        0       1990   
2       1371816893  40.495810  -74.196111        0       1970   
3       1371816915  28.812398  -80.883061        0       1987   
4       1371816917  44.959148  -85.884734        0       1955   
...            ...        ...         ...      ...        ...   
555714  1388534347  39.946837  -91.333331    

In [11]:
df_copy = df.copy()

# 2.) Select four columns to use as features (one just be trans_date_trans)

In [12]:
df_select = df[["trans_date_trans_time", "customer_classification", "amt", "city_pop", "is_fraud"]]

In [13]:
df_select.columns

Index(['trans_date_trans_time', 'customer_classification', 'amt', 'city_pop',
       'is_fraud'],
      dtype='object')

# 3.) Create your own variable out of trans_date. Create dummies for factor vars

In [19]:
type(df_select["trans_date_trans_time"][0])

str

In [14]:
df_select["trans_date_trans_time"] = pd.to_datetime(df_select["trans_date_trans_time"])

<ipython-input-14-99f721e4ce0f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_select["trans_date_trans_time"] = pd.to_datetime(df_select["trans_date_trans_time"])


In [21]:
dir(df_select["trans_date_trans_time"][0])

['__add__',
 '__array_priority__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__pyx_vtable__',
 '__radd__',
 '__reduce__',
 '__reduce_cython__',
 '__reduce_ex__',
 '__repr__',
 '__rsub__',
 '__setattr__',
 '__setstate__',
 '__setstate_cython__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__weakref__',
 '_date_repr',
 '_freq',
 '_freqstr',
 '_repr_base',
 '_round',
 '_set_freq',
 '_short_repr',
 '_time_repr',
 'asm8',
 'astimezone',
 'ceil',
 'combine',
 'ctime',
 'date',
 'day',
 'day_name',
 'day_of_week',
 'day_of_year',
 'dayofweek',
 'dayofyear',
 'days_in_month',
 'daysinmonth',
 'dst',
 'floor',
 'fold',
 'freq',
 'freqstr',
 'fromisocalendar',
 'fromisoformat',
 'fromordinal',
 'fromtimestamp',
 'hour',
 'is_leap_year',
 'is_month_end',
 'is_month_start',
 'is_

In [15]:
df_select["time_var"] = [i.hour for i in df_select["trans_date_trans_time"]]

<ipython-input-15-f84357779188>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_select["time_var"] = [i.hour for i in df_select["trans_date_trans_time"]]


In [16]:
X = pd.get_dummies(df_select, ["customer_classification"]).drop(["trans_date_trans_time", "is_fraud"], axis = 1)
y = df["is_fraud"]

In [24]:
X.head()

amt  city_pop  time_var  customer_classification_elderly  \
0   2.86    333497        12                                0   
1  29.84       302        12                                0   
2  41.28     34496        12                                0   
3  60.05     54767        12                                0   
4   3.19      1126        12                                1   

   customer_classification_middle_aged  customer_classification_young_adult  
0                                    1                                    0  
1                                    0                                    1  
2                                    1                                    0  
3                                    0                                    1  
4                                    0                                    0

In [17]:
resample_X = X
resample_y = y

# 5.) Train a Logistic regression.

In [18]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_normalized = scaler.fit_transform(resample_X)

In [19]:
from sklearn.linear_model import LogisticRegression

In [20]:
log_reg = LogisticRegression().fit(X_normalized, resample_y)

# 6.) The company you are working for wants to target at a False Positive rate of 5%. What threshold should you use? (Use oversampled data)

In [21]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_curve
import numpy as np

# train and fit logistic regression model
log_reg = LogisticRegression().fit(X_normalized, resample_y)

# predict class probabilities
probs = log_reg.predict_proba(X_normalized)[:, 1]

# calculate TPR and FPR for different threshold values
fpr, tpr, thresholds = roc_curve(resample_y, probs)

# target a specific false positive rate
target_fpr = 0.05

# find the threshold that corresponds to the target FPR
idx = np.where(fpr <= target_fpr)
threshold = thresholds[idx][-1]

# make predictions using the threshold
y_pred = probs >= threshold


In [22]:
print('Threshold:',threshold)

Threshold: 0.005278875377888082


If the company we are working for wants to target a False Positive rate of 5%, I found that we should use a 0.005278875377888082 threshold. Anything smaller than this results in a FP % greater than 5%, which is what we wanted to avoid. However, with the threshold this low, I believe we are adding a lot of potential bias to the model

# 7.) If the company makes .02*amt on True transactions and loses -amt on False (Use original data)

In [23]:
df_temp = df_select.copy()

In [85]:
df_temp.head()

trans_date_trans_time customer_classification    amt  city_pop  is_fraud  \
0   2020-06-21 12:14:25             middle_aged   2.86    333497         0   
1   2020-06-21 12:14:33             young_adult  29.84       302         0   
2   2020-06-21 12:14:53             middle_aged  41.28     34496         0   
3   2020-06-21 12:15:15             young_adult  60.05     54767         0   
4   2020-06-21 12:15:17                 elderly   3.19      1126         0   

   time_var  
0        12  
1        12  
2        12  
3        12  
4        12

In [24]:
df_temp["pred"] = log_reg.predict(resample_X)


/usr/local/lib/python3.8/dist-packages/sklearn/base.py:443: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(


In [25]:
df_temp = df_temp[["pred", "is_fraud", "amt"]]

In [26]:
df_temp.loc[df_temp['pred'] == 1].count()

pred        27991
is_fraud    27991
amt         27991
dtype: int64

In [27]:
df_temp.loc[df_temp['pred'] == 0].count()

pred        527728
is_fraud    527728
amt         527728
dtype: int64

In [29]:
df_temp.head()

pred  is_fraud    amt
0     0         0   2.86
1     0         0  29.84
2     0         0  41.28
3     0         0  60.05
4     0         0   3.19

In [28]:
df_tempc = df_temp.copy()

In [30]:
df_tempc['Altered_amt'] = df_tempc['amt'].copy()

In [31]:
df_tempc.loc[(df_tempc['pred'] == 0) & (df_tempc['is_fraud'] ==0), 'Altered_amt'] *= 1.2
df_tempc.loc[(df_tempc['pred'] == 1) & (df_tempc['is_fraud'] ==0), 'Altered_amt'] *= -1
df_tempc.loc[(df_tempc['pred'] == 1) & (df_tempc['is_fraud'] ==1), 'Altered_amt'] *= -1
df_tempc.loc[(df_tempc['pred'] == 0) & (df_tempc['is_fraud'] ==1), 'Altered_amt'] *= -1

In [32]:
df_tempc.head()

pred  is_fraud    amt  Altered_amt
0     0         0   2.86        3.432
1     0         0  29.84       35.808
2     0         0  41.28       49.536
3     0         0  60.05       72.060
4     0         0   3.19        3.828

In [33]:
round(df_tempc['Altered_amt'].sum(),2)

31393011.84

If the company makes .02*amt on True transactions and loses -amt on False, the company would make $31,393,011.84. This number was calculated on the assumption that we only make money if there's no fraud and we predict no fraud, while we lose the full amount in every other scenario.

# 8.) Using Logistic Regression Lasso to inform you. Would you use the selected features in a trusted prediction model?

In [36]:
# If most or all your variables go to 0 => Your data is garbage
# The regularization will tell us if our model has significance
# This of using coefficient strength similar to r^2
lasso_logreg = LogisticRegression(penalty='l1', solver='liblinear')
lasso_logreg.fit(X_normalized, resample_y)
coefs = lasso_logreg.coef_


In [44]:
coefs

array([[ 0.2841281 , -0.10558985,  0.185432  ,  0.13599199,  0.        ,
        -0.02016318]])

As seen above, the most of the coefficients are nonzero, which indicates that the model has significance and we could use the selected features in a trusted prediction model. 